Implementation of the networking using `libp2p` and Beaver triple generation using FFI

hdvanegasm opened the pull request:
# Network implementation using `libp2p`

This PR contains the implementation of the networking using `libp2p` library.

The implementation considers that the communication is performed in two ways: point-to-point and broadcast. For point-to-point communication, I used raw streams; for broadcast messages, I used gossipsub with only one topic.

Received point-to-point messages are passed to the calling function as a vector of bytes. On the other hand, received messages from the broadcast channel are passed to the calling functions using channels as the reception of the message is triggered by an event.

For point-to-point communication, the connected streams are stored in a `Arc<Mutex<HashMap<PeerId, Arc<Mutex<Stream>>>>>`. This allow to protect the list of connections from concurrent modifications, and also allows to protect each individual stream from concurrent read/write.

# Beaver triple generation using an FFI to `emp-tool` 

This PR also implements the generation of Beaver Triple in Rust by calling the thtfhe/triples C++ code components through FFI. The following is included in the PR:

- C Wrapper of CountNetIO and it's FFI implementation to be called from Rust
- Rust implementation of the Beaver Triple from thfhe/triples/test/triples.cpp
- End-to-end testing of beaver Triple generation
- Tracing added for logging info related to computational time for each component. It can be enabled using RUST_LOG=info
- Optional verification of triples is added as a feature verify. It can be build/tested using --features verify flag. The default does not do any verification
- (To be Removed): The generated Beaver triple is also stored in the .txt file for each parties.

<img width="982" height="615" alt="513139976-9029767b-5973-47ec-83e5-e30e8ad6065c" src="https://github.com/user-attachments/assets/5fe6de16-7e67-45d5-a24a-8dc7158b3ad0" />

hdvanegasm committed:
feat(emp-wrapper): added wrapper for OLEF2k.

hdvanegasm committed:
fix(emp-wrapper): fixed import and Cargo.toml

hdvanegasm committed:
fix(emp-wrapper): initial implementation of Rust side of the wrapper

hdvanegasm committed:
feat(emp-wrapper): implementation of the wrapper and bindings for OLE F2k.

This commit contains the complete implementation of the wrapper and bindings for OLE F2k without documentation. It compiles, although I haven't tested it yet.

hdvanegasm committed:
feat(emp-wrapper): implementation of the wrapper for OLE Z2k.

The implemented code has correct linkage to the C++ library. Some tests are needed to check that all the objects are being connected safely.

hdvanegasm committed:
Merge remote-tracking branch 'origin/main' into emp-bindings

# Conflicts:
#	Cargo.toml

hdvanegasm committed:
fix(emp-wrapper): remove clippy warnings.

hdvanegasm committed:
fix(emp-wrapper): remove blocking test.

hdvanegasm committed:
feat(p2p): implemented networking stack using `libp2p`

hdvanegasm committed:
fix(p2p): remove stream when the connection is closed.

hdvanegasm committed:
test(p2p): WIP: started tests to the network

hdvanegasm committed:
fix(emp-bindings): added Cargo.toml file

hdvanegasm committed:
fix(emp-bindings): added lib.rs file

hdvanegasm committed:
fix(emp-bindings): added lib.rs file

hdvanegasm committed:
feat(emp-bindings): added a README.md file with the requirements

hdvanegasm committed:
fix(emp-wrapper): added missing files

hdvanegasm committed:
fix(emp-wrapper): remove .idea

hdvanegasm committed:
fix(emp-wrapper): fix the imports.

hdvanegasm committed:
feat(emp-bindings): fix files.

hdvanegasm committed:
feat(p2p): refractor architecture in the network.

hdvanegasm committed:
feat(p2p): added mechanisms to open raw strams.

hdvanegasm committed:
feat(p2p): added mechanisms to open retry the dialing.

man2706kum committed:
feat(emp-wrapper): added wrapper for CountNetIO

man2706kum committed:
feat(beaver-triple): started with beaver triple generation

man2706kum committed:
feat(beaver-triple): beaver triple generation added using NetIo

man2706kum committed:
fix(beaver-triple): beaver triple uses CountNetIo instead of NetIo.

all related ffi changed accordingly as per original c code.

Test code for beaver triple added(not working).

man2706kum committed:
fix(CountNetIo): CountNetIo initialization

man2706kum committed:
fix(beaver-triple): switched back to using NetIo instead of CountNetIo

NetIo ffi fixed, allows null ptr

all ffi changed to use NetIo

hdvanegasm committed:
feat(gitignore): add .idea folder

man2706kum committed:
fix(beaver-triple): fixed undefined symbol: BN_new

FerretCot component working

need to set LD_LIBRARY_PATH manually using export

man2706kum committed:
Merge branch 'emp-bindings' of https://github.com/hashcloak/Ajax into emp-bindings

man2706kum committed:
fix(beaver-triple): triple generation end to end working

triple also being stored in files(to be removed)

man2706kum committed:
feat(beaver-triple): switched to using CountNetIO instead of NetIO

man2706kum committed:
correctness testing added as a feature

can be tested with  flag

man2706kum committed:
feat(veaver-triple): tracing added for triple generation

hdvanegasm committed:
feat(gitignore): add .idea folder

hdvanegasm committed:
feat(cargo): update cargo.

hdvanegasm committed:
fix(p2p): fix some deadlocks.

The previous version of the implementation contained some deadlocks given that we were for the swarm to dial and open streams, but this blocked the progress as the swarm wasn't polled again, resulting in a deadlock.

hdvanegasm committed:
fix(p2p): fix some deadlocks.

hdvanegasm committed:
fix(p2p): fix all the deadlocks and correct network errors.

hdvanegasm committed:
feat(p2p): simplify node building and started the use of P2P network in DN07

hdvanegasm committed:
feat(p2p): removed unused imports

hdvanegasm committed:
fix(p2p): add synchronization to the network and correct minor issues.

hdvanegasm committed:
Merge pull request #5 from hashcloak/p2p

Implementation of P2P library

hdvanegasm committed:
fix(p2p): fix clippy suggestions.

hdvanegasm committed:
docs(p2p): added missing documentation.

hdvanegasm committed:
Merge pull request #6 from hashcloak/p2p

P2P documentation

hdvanegasm committed:
Merge branch 'emp-bindings' into p2p

# Conflicts:
#	emp-bindings/Cargo.toml
#	emp-bindings/src/lib.rs
#	thfhe/triples/CMakeLists.txt
#	thfhe/triples/include/wrapper/io.h

hdvanegasm committed:
Merge pull request #7 from hashcloak/p2p

Merge with `emp-bindings`

hdvanegasm committed:
fix(p2p): update github actions and build system.

hdvanegasm committed:
fix(p2p): small test fix

hdvanegasm committed:
fix(p2p): small test fix

hdvanegasm committed:
ci: fixed workflows to compile the wrapper.

hdvanegasm committed:
ci: fixed underlying OS.

hdvanegasm committed:
ci: fixed upload artifact version.

hdvanegasm committed:
Merge pull request #8 from hashcloak/p2p

Update CI

man2706kum committed:
fix(beaver-triple): NetIO code removed

ios guard and cots guard dropped manually after use

man2706kum committed:
feat(beaver-triple): benchmark skeleton

man2706kum committed:
Merge branch 'primus-labs:p2p' into p2p

man2706kum committed:
fix(beaver-triple): potential deadlock fixed

io and cot locks released early to prevent deadlock

man2706kum committed:
Merge branch 'p2p' of https://github.com/hashcloak/Ajax into p2p

man2706kum committed:
fix(beaver-triple): io header file removed and tracing info modified

hdvanegasm committed:
Merge pull request #9 from hashcloak/p2p

Triple Generation Benchmarking

hdvanegasm commented:
# Benchmarking

In this PR, we also added benchmarks for the triple generation. In the following images you will see the results:

## Rust implementation using the FFI  - Debug

<img width="1280" height="605" alt="image" src="https://github.com/user-attachments/assets/f24a6cd9-1482-40b9-84a9-11b994c505c5" />

## Rust implementation using the FFI  - Release

<img width="1280" height="610" alt="image" src="https://github.com/user-attachments/assets/6b4bafe6-a59d-45a6-a9c3-bfb92022190e" />

## C++ implementation using EMP-tool

<img width="1280" height="542" alt="image" src="https://github.com/user-attachments/assets/f46b83c0-edf7-4161-8a3e-a0214baddd0f" />

haofeiliang commented:
Is there any relationship between the ".idea" folder and the project?  
If not, can it be removed from the commit?

haofeiliang commented on thfhe/src/distdec.rs:
```diff
@@ -53,10 +53,10 @@ where
         let a_mul_s_sum = backend.inner_product_additive_const_p(ax, &additive_shares);
 
-        // m_add_e = b-as mod q
+        // m_add_e = b-as p2p q
```
Here, `mod` means modular arithmetic.

haofeiliang replied:
There are many such unexpected replacements, and they need to be fixed.

haofeiliang commented on thfhe/src/secret_key.rs:
```diff
@@ -135,12 +135,12 @@ where
 }
 
-/// Tonelli-Shanks 算法 (适用于 p ≡ 1 mod 4)
+/// Tonelli-Shanks 算法 (适用于 p ≡ 1 p2p 4)
```
This also needs to be fixed.

haofeiliang reviewed (changes requested):
This is a big code change, but it seems to be mainly about network optimization and Beaver triple generation. I didn’t go through all the implementation details carefully (sorry, I’m not very familiar with p2p and FFI). However, I think as long as all the `.idea` folders are removed and the incorrect `mod` → `p2p` replacements are fixed, it should be ready to merge.

haofeiliang commented:
In addition, I noticed that the `p2p` implementation mainly uses `tokio::sync::Mutex`. Can these be replaced with `Mutex` from `std` or `parking_lot`? Also, the maps are using `Arc<Mutex<HashMap>>`; could this be replaced with a `DashMap` or a `HashMap` from the `scc` crate?

hdvanegasm committed:
fix: remove .idea folder

hdvanegasm committed:
fix: remove .idea folder

hdvanegasm committed:
fix: remove .idea folder

hdvanegasm committed:
fix(mpc): change word "p2p" for mode to fix comments.

hdvanegasm commented:
> In addition, I noticed that the `p2p` implementation mainly uses `tokio::sync::Mutex`. Can these be replaced with `Mutex` from `std` or `parking_lot`? Also, the maps are using `Arc<Mutex<HashMap>>`; could this be replaced with a `DashMap` or a `HashMap` from the `scc` crate?

I changed the `Mutex<HashMap>` for `DashMap`.

Regarding the change of `Mutex`, I think that in this case is not recommended to use the `std::sync::Mutex` because libp2p works in the `tokio` runtime. Hence, if we use `std::sync::Mutex`, a call to `lock()` would block the thread completely and also block the runtime execution preventing it from polling tasks from the pool. I would recommend to stay with `tokio::sync::Mutex`. Do you agree with this argument @serendipity-crypto ? What are the reasons behind not using `tokio::sync::Mutex`?

hdvanegasm committed:
fix(p2p): change `Mutex<HashMap>` for `DashMap`

hdvanegasm committed:
Merge pull request #10 from hashcloak/p2p

Addressing comments to the first revision

haofeiliang commented:
> > In addition, I noticed that the `p2p` implementation mainly uses `tokio::sync::Mutex`. Can these be replaced with `Mutex` from `std` or `parking_lot`? Also, the maps are using `Arc<Mutex<HashMap>>`; could this be replaced with a `DashMap` or a `HashMap` from the `scc` crate?
> 
> I changed the `Mutex<HashMap>` for `DashMap`.
> 
> Regarding the change of `Mutex`, I think that in this case is not recommended to use the `std::sync::Mutex` because libp2p works in the `tokio` runtime. Hence, if we use `std::sync::Mutex`, a call to `lock()` would block the thread completely and also block the runtime execution preventing it from polling tasks from the pool. I would recommend to stay with `tokio::sync::Mutex`. Do you agree with this argument @serendipity-crypto ? What are the reasons behind not using `tokio::sync::Mutex`?

I agree that `lock()` may block the thread and runtime. Let's keep using `tokio::sync::Mutex`.

haofeiliang reviewed (approved):
LGTM. After @YounesTal1 reviews the PR, if everything looks good, we can go ahead and merge it.

hdvanegasm commented:
An additional question @serendipity-crypto : Do you agree with the new GitHub workflow? I removed `clippy` as it will always fail given that there are some Rust style issues but I supposed that that's not relevant at this point. What do you think about it?

haofeiliang commented:
> An additional question @serendipity-crypto : Do you agree with the new GitHub workflow? I removed `clippy` as it will always fail given that there are some Rust style issues but I supposed that that's not relevant at this point. What do you think about it?

This is fine. The previous `clippy` settings were a bit too strict; it might be better to adjust the `clippy` rules, though that’s not related to this PR. If needed, we can update them in another PR later. For now, removing it is totally fine.

haofeiliang committed:
Change the branch of the dependencies from `primus-fhe` (#12)

YounesTal1 commented on mpc/tests/dn.rs:
```diff
@@ -1,117 +1,188 @@
-    const THRESHOLD: u32 = 1;
-    const BASE_PORT: u32 = 50000;
+#[tokio::test]
+async fn test_secret_sharing_and_recovery() {
```
This test doesn't end when i set `NUM_PARTIES` to 5 and `THRESHOLD` to 2 (I waited for minutes)

hdvanegasm replied:
I think that the problem is in this section:

https://github.com/primus-labs/Ajax/blob/341c090f78dc78cd53c7142248a4cefb19cda3c0/mpc/tests/dn.rs#L94-L103

I think that this section is not correct because only Party 0 and Party 1 are interacting, but they are constructing a polynomial of degree 2, hence, they need 3 valid shares to be able to reconstruct the secret, but there are only two parties with shares. @serendipity-crypto Could you confirm if that's the actual issue? I don't want to make wrong claims on the existing code.

haofeiliang replied:
This old test is for two parties to reconstruct the secret. It can be deleted.

hdvanegasm replied:
I will delete the test. Thanks!

YounesTal1 commented on network/tests/p2p.rs:
```diff
@@ -0,0 +1,539 @@
+
+#[tokio::test]
+#[serial]
+async fn send_and_receive() {
```
This test fails when i set `NUM_PARTIES` to 35

hdvanegasm replied:
This test is working for me with `NUM_PARTIES = 35`. Might it be a race condition? I will keep checking with other values. It passes for me even with `NUM_PARTIES=100`

hdvanegasm replied:
I see the error now. I think that it's a `libp2p` inconvenience. When you run a node and it takes a long time to connect to it, then `libp2p` disconnects the node. I will see how can I solve this issue.

hdvanegasm replied:
@YounesTal1 Could you please pull and then try this test again in your computer?

YounesTal1 replied:
Ah it is still failing, after 15min i think.

hdvanegasm replied:
I managed to find the source of the issue. The problem is that Unix-like system has a default maximum limit of 1024 file descriptors per process. In this case, given that each party needs to open a new stream with all the other parties, then you need to consume at least 35 * 34 = 1190 file descriptors (plus additional file descriptors for other network functionalities) which surpasses the bound of 1024 mentioned before. The solution is that you increase the number of file descriptors using the command `ulimit -n <new_limit>`. That should solve the error. At least it solved the error in my case.

Please increase the file descriptors in your system and try again. Let me know if the issue is still present.

YounesTal1 replied:
It is working now with increasing the `ulimit`, thanks!

YounesTal1 reviewed (commented):
I had a look over the code. I left two comments. Thanks

hdvanegasm committed:
fix(p2p): added network mechanisms for slower contexts.

The network mechanisms are: a pinging protocol to avoid disconnection because of idle nodes, and an increase of retrials for slower executions.

hdvanegasm committed:
Merge pull request #1 from hashcloak/triples-tfhe-integration

fix(p2p): added network mechanisms for slower contexts.

hdvanegasm committed:
fix(mpc): rollback previous values of test.

hdvanegasm committed:
Merge pull request #13 from hashcloak/p2p

Corrected issues from revision of https://github.com/primus-labs/Ajax/pull/3

hdvanegasm committed:
feat(p2p): added tracing for tests.

hdvanegasm committed:
Merge pull request #14 from hashcloak/p2p

feat(p2p): added tracing for tests.

hdvanegasm committed:
fix(p2p): added tracing for tests.

hdvanegasm committed:
Merge pull request #15 from hashcloak/p2p

fix(p2p): added tracing for tests.

hdvanegasm committed:
fix(mpc): fixed test for sharing and recovering.

hdvanegasm committed:
fix(mpc): automated array generation in test.

hdvanegasm committed:
Merge pull request #16 from hashcloak/p2p

Fix test
